common/flatpak-instance.c

```diff
@@ -0,0 +1,317 @@
+  char     *branch;
+  char     *commit;
+  char     *runtime;
+  char     *runtime_commit;
+
+  int       pid;
+  int       child_pid;
+};
+
+G_DEFINE_TYPE_WITH_PRIVATE (FlatpakInstance, flatpak_instance, G_TYPE_OBJECT)
+
+static void
+flatpak_instance_finalize (GObject *object)
+{
+  FlatpakInstance *self = FLATPAK_INSTANCE (object);
+  FlatpakInstancePrivate *priv = flatpak_instance_get_instance_private (self);
+
+  g_free (priv->id);
+  g_free (priv->dir);
+  g_free (priv->app);
+  g_free (priv->arch);
+  g_free (priv->branch);
+  g_free (priv->commit);
+  g_free (priv->runtime);
+  g_free (priv->runtime_commit);
+
+  if (priv->info)
+    g_key_file_unref (priv->info);
+
+  G_OBJECT_CLASS (flatpak_instance_parent_class)->finalize (object);
+}
+
+static void
+flatpak_instance_class_init (FlatpakInstanceClass *klass)
+{
+  GObjectClass *object_class = G_OBJECT_CLASS (klass);
+
+  object_class->finalize = flatpak_instance_finalize;
+}
+
+static void
+flatpak_instance_init (FlatpakInstance *self)
+{
+}
+
+const char *
+flatpak_instance_get_id (FlatpakInstance *self)
+{
+  FlatpakInstancePrivate *priv = flatpak_instance_get_instance_private (self);
+
+  return priv->id;
+}
+
+const char *
+flatpak_instance_get_app (FlatpakInstance *self)
+{
+  FlatpakInstancePrivate *priv = flatpak_instance_get_instance_private (self);
+
+  return priv->app;
+}
+
+const char *
+flatpak_instance_get_arch (FlatpakInstance *self)
+{
+  FlatpakInstancePrivate *priv = flatpak_instance_get_instance_private (self);
+
+  return priv->arch;
+}
+
+const char *
+flatpak_instance_get_branch (FlatpakInstance *self)
+{
+  FlatpakInstancePrivate *priv = flatpak_instance_get_instance_private (self);
+
+  return priv->branch;
+}
+
+const char *
+flatpak_instance_get_commit (FlatpakInstance *self)
+{
+  FlatpakInstancePrivate *priv = flatpak_instance_get_instance_private (self);
+
+  return priv->commit;
+}
+
+const char *
+flatpak_instance_get_runtime (FlatpakInstance *self)
+{
+  FlatpakInstancePrivate *priv = flatpak_instance_get_instance_private (self);
+
+  return priv->runtime;
+}
+
+const char *
+flatpak_instance_get_runtime_commit (FlatpakInstance *self)
+{
+  FlatpakInstancePrivate *priv = flatpak_instance_get_instance_private (self);
+
+  return priv->runtime_commit;
+}
+
+int
+flatpak_instance_get_pid (FlatpakInstance *self)
+{
+  FlatpakInstancePrivate *priv = flatpak_instance_get_instance_private (self);
+
+  return priv->pid;
+}
+
+int
+flatpak_instance_get_child_pid (FlatpakInstance *self)
+{
+  FlatpakInstancePrivate *priv = flatpak_instance_get_instance_private (self);
+
+  return priv->child_pid;
+}
+
+GKeyFile *
+flatpak_instance_get_info (FlatpakInstance *self)
+{
+  FlatpakInstancePrivate *priv = flatpak_instance_get_instance_private (self);
+
+  return priv->info;
+}
+
+gboolean
+flatpak_instance_is_running (FlatpakInstance *self)
+{
+  FlatpakInstancePrivate *priv = flatpak_instance_get_instance_private (self);
+
+  flatpak_run_gc_ids ();
+
```

Comment:
This is a bit overkill for _is_running(). It will enumerate all processes and stuff. All you have to do is to try to fcntl F_GETLK the .ref file in the directory, and it if returns l.l_type == F_UNLCK (or doesn't exist) then its not running.